File: app/src/main/java/com/example/trigger/https/CertificatePreference.java
package com.example.trigger.https;

import android.content.Context;
import android.content.Intent;
import android.preference.Preference;
import android.preference.SwitchPreference;
import android.util.AttributeSet;

import java.security.cert.Certificate;

import com.example.trigger.SetupActivity;


/*
* Stores a string in preference. Indicated the presence of data with a switch.
* On click, the file chooser opens.
*/
public class CertificatePreference extends SwitchPreference {
    private Certificate certificate;
    private Context context;
    static CertificatePreference self;

    public CertificatePreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        this.context = context;
        final CertificatePreference self = this;

        // tell the superclass that we handle the value on out own!
        setPersistent(false);

        if (this.certificate == null) {
            setChecked(false);
        } else {
            setChecked(true);
        }

        this.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                if (self.certificate == null) {
                    self.setChecked(false);
                } else {
                    self.setChecked(true);
                }
                
                return false;
            }
        });

        this.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                String register_url = ((SetupActivity) self.context).getRegisterUrl();

                // store in public static member - hack!
                CertificatePreference.self = self;

                Intent intent = new Intent(self.context, CertificateActivity.class);
                intent.putExtra("register_url", register_url);

                self.context.startActivity(intent);
                return false;
            }
        });
    }

    public void setCertificate(Certificate certificate) {
        this.certificate = certificate;

        if (this.certificate == null) {
            setChecked(false);
        } else {
            setChecked(true);
        }
    }

    public Certificate getCertificate() {
        return certificate;
    }

    @Override
    protected void onSetInitialValue(boolean restorePersistedValue, Object defaultValue) {
        if (certificate == null) {
            super.onSetInitialValue(restorePersistedValue, false);
        } else {
            super.onSetInitialValue(restorePersistedValue, true);
        }
    }
}

File: app/src/main/java/app/trigger/https/CertificatePreference.kt
package app.trigger.https

import android.content.Intent
import android.preference.SwitchPreference
import android.preference.Preference.OnPreferenceChangeListener
import android.preference.Preference
import android.preference.Preference.OnPreferenceClickListener
import app.trigger.SetupActivity
import android.content.Context
import android.util.AttributeSet
import java.security.cert.Certificate

/*
* Stores a string in preference. Indicated the presence of data with a switch.
* On click, the file chooser opens.
*/
class CertificatePreference(private val context_: Context, attrs: AttributeSet?) : SwitchPreference(context_, attrs) {
    public var certificate: Certificate? = null
        set(value) {
            field = value
            isChecked = (certificate != null)
        }

    override fun onSetInitialValue(restorePersistedValue: Boolean, defaultValue: Any) {
        if (certificate == null) {
            super.onSetInitialValue(restorePersistedValue, false)
        } else {
            super.onSetInitialValue(restorePersistedValue, true)
        }
    }

    companion object {
        var self: CertificatePreference? = null
    }

    init {
        val self = this

        // tell the superclass that we handle the value on out own!
        isPersistent = false
        isChecked = (certificate != null)
        this.onPreferenceChangeListener = OnPreferenceChangeListener { preference: Preference?, newValue: Any? ->
            self.isChecked = (self.certificate != null)
            false
        }
        this.onPreferenceClickListener = OnPreferenceClickListener { preference: Preference? ->
            val register_url = (self.context as SetupActivity).getRegisterUrl()

            // store in public static member - hack!
            Companion.self = self
            val intent = Intent(self.context, CertificateActivity::class.java)
            intent.putExtra("register_url", register_url)
            intent.putExtra("certificate_preference_title", self.title.toString())
            self.context.startActivity(intent)
            false
        }
    }
}